### PR TITLE
Add adaptive rubrics support with RubricVerifier and utility functions

### PR DESCRIPTION
**Pull request by RulinShao**
### Summary
This PR migrates the evolving rubrics (formerly "adaptive rubrics") feature to enable dynamic rubric generation during RL training. Evolving rubrics are criteria generated on-the-fly based on model responses to better distinguish high-quality from low-quality answers.

### Key Changes

**New Module: `open_instruct/rubrics/`**
- `rubric_utils.py` - Core functions for evolving rubric generation and management
- `run_utils.py` - LLM API utilities (LiteLLM) for rubric judging
- `test_rubrics.py` - Unit tests for the rubrics module

**Configuration (`data_loader.py`)**
New `StreamingDataLoaderConfig` parameters:
- `apply_evolving_rubric_reward: bool` - Enable evolving rubric rewards
- `use_rubric_buffer: bool` - Maintain rubric buffer across training steps  
- `max_active_rubrics: int` - Max evolving rubrics per query (default: 5)
- `cache_evolving_rubric_data_dir: str | None` - Directory to cache rubric data

**Verifier (`ground_truth_utils.py`)**
- Added `RubricVerifier` class for scoring responses against rubrics via LLM judge

**Test Script**
- `scripts/train/debug/evolving_rubric_mini_test.sh` - Single-GPU E2E test

### File Structure
```
open_instruct/rubrics/
├── __init__.py
├── rubric_utils.py    # Core evolving rubric functions
├── run_utils.py       # LiteLLM utilities  
└── test_rubrics.py    # Unit tests
```

### Breaking Changes
- Renamed `adaptive_rubric` → `evolving_rubric` in all APIs and configs
- Removed deprecated flags: `save_adaptive_rubrics`, `normalize_rubric_scores`, `use_static_rubrics_as_persistent_rubrics`

### Testing
```bash
# Unit tests (23 pass, 6 skip without API key)
uv run pytest open_instruct/rubrics/test_rubrics.py -v

# E2E test (requires GPU + API key)
export OPENAI_API_KEY=your_key
bash scripts/train/debug/evolving_rubric_mini_test.sh
```

### CODE REVIEW

File: open_instruct/search_rewards/utils/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Utils package for search rewards
```
Comment (hamishivi) [inline open_instruct/search_rewards/utils/__init__.py]: Could we move this into a folder called rubrics or similar?

File: open_instruct/data_loader.py [hunk 2]
```diff
@@ -362,6 +362,24 @@ class StreamingDataLoaderConfig:
...
+    """Whether to normalize rubric scores within each rollout group."""
+    use_rubric_buffer: bool = False
+    """Whether to maintain a buffer of rubrics across training steps."""
+    max_active_rubrics: int = 5
+    """Maximum number of active adaptive rubrics per query."""
+    use_static_rubrics_as_persistent_rubrics: bool = True
+    """Whether to use static/ground-truth rubrics as persistent rubrics in the buffer."""
+    cache_adaptive_rubric_data_dir: str | None = None
+    """Directory to cache adaptive rubric generation data for debugging/analysis."""
+    save_adaptive_rubrics: bool = False
```
Comment (hamishivi) [inline open_instruct/data_loader.py]: perhaps we can just make it if cache_adaptive_rubric_data_dir is set, we assume the user wants to save?

File: open_instruct/data_loader.py [hunk 3]
```diff
@@ -362,6 +362,24 @@ class StreamingDataLoaderConfig:
...
+    """Whether to generate and apply adaptive rubrics for reward computation."""
+    use_full_responses_for_adaptive_rubric: bool = True
+    """Whether to use full responses (vs extracted answers) for adaptive rubric generation."""
+    normalize_rubric_scores: bool = False
+    """Whether to normalize rubric scores within each rollout group."""
+    use_rubric_buffer: bool = False
+    """Whether to maintain a buffer of rubrics across training steps."""
+    max_active_rubrics: int = 5
+    """Maximum number of active adaptive rubrics per query."""
+    use_static_rubrics_as_persistent_rubrics: bool = True
```
Comment (hamishivi) [inline open_instruct/data_loader.py]: did we ever turn this off? can we remove this option?

File: open_instruct/data_loader.py [hunk 4]
```diff
@@ -362,6 +362,24 @@ class StreamingDataLoaderConfig:
     non_stop_penalty: bool = False
     non_stop_penalty_value: float = 0.0
 
+    # Adaptive rubric reward
```
Comment (hamishivi) [inline open_instruct/data_loader.py]: In general itd be useful to make these args as minimal as possible

File: open_instruct/ground_truth_utils.py [hunk 5]
```diff
@@ -922,6 +922,120 @@ def get_config_class(cls) -> type:
...
+
+    def __init__(self, verifier_config: RubricVerifierConfig) -> None:
+        super().__init__("rubric", verifier_config=verifier_config, weight=1.0)
+        self.config = verifier_config
+
+    async def async_call(
+        self, tokenized_prediction: list[int], prediction: str, label: Any, query: str | None = None
+    ) -> VerificationResult:
+        """Score response against all rubrics in the ground truth."""
+        from open_instruct.search_rewards.utils.run_utils import (  # noqa: PLC0415
```
Comment (hamishivi) [inline open_instruct/ground_truth_utils.py]: nit: move imports to top of file

File: open_instruct/ground_truth_utils.py [hunk 6]
```diff
@@ -922,6 +922,120 @@ def get_config_class(cls) -> type:
...
+            extract_json_from_response,
+            run_litellm_async,
+        )
+
+        # Parse the ground truth
```
Comment (hamishivi) [inline open_instruct/ground_truth_utils.py]: unnecessary comment

File: .gitignore [hunk 7]
```diff
@@ -160,3 +160,10 @@ dmypy.json
 cache/
 local_dataset_cache/
 scratch/
+
+open_instruct/rl-rag-eval/
```
Comment (hamishivi) [inline .gitignore]: can we remove these items from the gitignore?

File: open_instruct/search_rewards/utils/test_all.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
```
Comment (hamishivi) [inline open_instruct/search_rewards/utils/test_all.py]: instead of this, can you make the tests just use unittest matching the tests in the rest of the repo?

File: open_instruct/search_rewards/utils/__init__.py (hunk 1 above)
Comment (RulinShao) [reply open_instruct/search_rewards/utils/__init__.py]: fixed

File: open_instruct/data_loader.py (hunk 2 above)
Comment (RulinShao) [reply open_instruct/data_loader.py]: fixed

File: open_instruct/data_loader.py (hunk 3 above)
Comment (RulinShao) [reply open_instruct/data_loader.py]: was using them in my local tests/ablations, but not in the final paper; just cleaned up these unused arguments

File: open_instruct/data_loader.py (hunk 4 above)
Comment (RulinShao) [reply open_instruct/data_loader.py]: cleaned

File: open_instruct/ground_truth_utils.py (hunk 5 above)
Comment (RulinShao) [reply open_instruct/ground_truth_utils.py]: fixed

File: open_instruct/ground_truth_utils.py (hunk 6 above)
Comment (RulinShao) [reply open_instruct/ground_truth_utils.py]: fixed

File: .gitignore (hunk 7 above)
Comment (RulinShao) [reply .gitignore]: fixed

File: open_instruct/search_rewards/utils/test_all.py (hunk 8 above)
Comment (RulinShao) [reply open_instruct/search_rewards/utils/test_all.py]: fixed

File: open_instruct/ground_truth_utils.py [hunk 9]
```diff
@@ -922,6 +923,114 @@ def get_config_class(cls) -> type:
...
+
+        # Score each rubric in parallel
+        async def score_rubric(rubric: dict) -> tuple[float, float]:
+            description = rubric.get("description") or rubric.get("rubric_item") or rubric.get("Ingredient", "")
+            weight = rubric.get("weight", 1.0)
+
+            user_prompt = f"<question>{question}</question>\n<response>{prediction}</response>\n<criterion>{description}</criterion>"
+
+            try:
+                model_name = os.environ.get("RUBRIC_JUDGE_MODEL", self.config.rubric_judge_model)
```
Comment (hamishivi) [inline open_instruct/ground_truth_utils.py:991]: can we add the env vars used into the docstring?

File: open_instruct/ground_truth_utils.py [hunk 10]
```diff
@@ -922,6 +923,114 @@ def get_config_class(cls) -> type:
...
+                resp = await run_litellm_async(
+                    model_name=model_name,
+                    system_prompt=self.RUBRIC_SCORING_PROMPT,
```
Comment (hamishivi) [inline open_instruct/ground_truth_utils.py]: maybe make this a constant inside the rubric folder?

Comment (RulinShao) [reply open_instruct/ground_truth_utils.py]: fixed, also fixed it for rubric generator

File: open_instruct/ground_truth_utils.py (hunk 9 above)
Comment (RulinShao) [reply open_instruct/ground_truth_utils.py:991]: fixed, did the same for rubric generation prompt too

File: open_instruct/rubrics/run_utils.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+            except json.JSONDecodeError:
+                try:
+                    # Fix doubled braces (e.g., '{{' -> '{', '}}' -> '}')
+                    fixed_braces = json_str.replace("{{", "{").replace("}}", "}")
+                    return json.loads(fixed_braces)
+                except json.JSONDecodeError:
+                    # Try next { position
+                    json_start = response.find("{", json_start + 1)
+                    continue
+        break
```
[the remark this answers is not in the document]
Comment (RulinShao) [reply open_instruct/rubrics/run_utils.py]: fixed

File: open_instruct/ground_truth_utils.py [hunk 12]
```diff
@@ -922,6 +923,114 @@ def get_config_class(cls) -> type:
...
+        question = label.get("query") or label.get("Question") or query
+        rubrics = label.get("rubrics", [])
+
+        if not rubrics:
+            logger.warning("No rubrics found in ground truth")
+            return VerificationResult(score=0.0)
+
+        # Score each rubric in parallel
+        async def score_rubric(rubric: dict) -> tuple[float, float]:
+            description = rubric.get("description") or rubric.get("rubric_item") or rubric.get("Ingredient", "")
```
[the remark this answers is not in the document]
Comment (RulinShao) [reply open_instruct/ground_truth_utils.py:981]: fixed

File: open_instruct/data_loader.py [hunk 13]
```diff
@@ -362,6 +362,16 @@ class StreamingDataLoaderConfig:
     non_stop_penalty: bool = False
     non_stop_penalty_value: float = 0.0
 
+    # Evolving rubric reward
+    apply_evolving_rubric_reward: bool = False
+    """Whether to generate and apply evolving rubrics for reward computation."""
+    use_rubric_buffer: bool = False
```
Comment (hamishivi) [inline open_instruct/data_loader.py]: Do we ever want this to be false but RLER is true?

File: open_instruct/rubrics/__init__.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    RUBRIC_GENERATION_MODEL: LLM model for generating new rubrics dynamically.
+        Default: "gpt-4.1"
+    LITELLM_MAX_CONCURRENT_CALLS: Maximum concurrent LiteLLM API calls.
+        Default: "256"
+    LITELLM_DEFAULT_TIMEOUT: Default timeout for LiteLLM calls in seconds.
+        Default: "600"
+"""
+
+# Prompt used by RubricVerifier to score responses against rubrics
+RUBRIC_SCORING_PROMPT = """You will be given a question someone asked (in <question></question> tags) and the corresponding response (in <response></response> tags) given to them by an assistant. You will then be given a specific criterion of the response to evaluate (in <criterion></criterion> tags).
```
Comment (hamishivi) [inline open_instruct/rubrics/__init__.py]: maybe put these in a prompts.py or constants.py instead of init

**Review by hamishivi — CHANGES_REQUESTED** [review]

File: open_instruct/data_loader.py (hunk 13 above)
Comment (RulinShao) [reply open_instruct/data_loader.py]: oh not really, just removed it!

File: open_instruct/rubrics/__init__.py (hunk 14 above)
Comment (RulinShao) [reply open_instruct/rubrics/__init__.py]: fixed

File: scripts/train/debug/evolving_rubric_mini_test.sh [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    --vllm_sync_backend gloo \
+    --vllm_tensor_parallel_size 1 \
+    --vllm_enforce_eager \
+    --lr_scheduler_type constant \
+    --apply_verifiable_reward true \
+    --seed 1 \
+    --save_freq 25 \
+    --try_launch_beaker_eval_jobs_on_weka False \
+    --gradient_checkpointing \
+    --max_tool_calls 0 \
```
Comment (hamishivi) [inline scripts/train/debug/evolving_rubric_mini_test.sh:106]:
```suggestion
    --max_tool_calls 0 \
    --remap_verifier general_rubric=rubric \
```

**Review by hamishivi — CHANGES_REQUESTED** [review]
works for me! one fix to script + would be good to add some basic metrics on the rubrics.

File: scripts/train/debug/evolving_rubric_mini_test.sh (hunk 15 above)
Comment (RulinShao) [reply scripts/train/debug/evolving_rubric_mini_test.sh:106]: fixed

File: open_instruct/rubrics/metrics.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
+"""Global metrics for evolving rubrics.
+
+Provides functions to compute and log:
+- Average reward from evolving rubrics vs persistent (static) rubrics.
+- Number of newly generated evolving rubrics per step.
+- Number of active rubrics in the buffer.
+- Rubric buffer filtering (deactivating uninformative rubrics).
+"""
+
+from __future__ import annotations
```
Comment (hamishivi) [inline open_instruct/rubrics/metrics.py]: this isnt needed, I think

File: scripts/train/debug/evolving_rubric_mini_test.sh [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+echo "=============================================="
+echo "Starting Evolving Rubric Mini Training Test"
+echo "=============================================="
+echo "Model: ${model_path}"
+echo "Dataset: ${dataset_list}"
+echo "Exp name: ${exp_name}"
+echo "Repo root: ${REPO_ROOT}"
+echo "=============================================="
+
+python3 open_instruct/grpo_fast.py \
```
Comment (hamishivi) [inline scripts/train/debug/evolving_rubric_mini_test.sh]:
```suggestion
uv run python open_instruct/grpo_fast.py \
```

We prefer to use uv for all dep management
